sdk/servicebus/service-bus/src/core/batchingReceiver.ts

```diff
@@ -325,200 +329,202 @@ export class BatchingReceiverLite {
+      }
 
-      // Action to be performed on the "receiver_drained" event.
-      const onReceiveDrain: OnAmqpEvent = () => {
-        receiver.removeListener(ReceiverEvents.receiverDrained, onReceiveDrain);
-        receiver.drain = false;
+      reject(translateServiceBusError(error));
+    };
 
+    let abortSignalCleanupFunction: (() => void) | undefined = undefined;
+
+    // Final action to be performed after
+    // - maxMessageCount is reached or
+    // - maxWaitTime is passed or
+    // - newMessageWaitTimeoutInSeconds is passed since the last message was received
+    const finalAction = (): void => {
+      // Drain any pending credits.
+      if (receiver.isOpen() && receiver.credit > 0) {
+        logger.verbose(`${loggingPrefix} Draining leftover credits(${receiver.credit}).`);
+
+        // setting .drain and combining it with .addCredit results in (eventually) sending
+        // a drain request to Service Bus. When the drain completes rhea will call `onReceiveDrain`
+        // at which point we'll wrap everything up and resolve the promise.
+        receiver.drain = true;
+        receiver.addCredit(1);
+      } else {
         logger.verbose(
-          `${loggingPrefix} Drained, resolving receiveMessages() with ${brokeredMessages.length} messages.`
+          `${loggingPrefix} Resolving receiveMessages() with ${brokeredMessages.length} messages.`
         );
+        resolveImmediately(brokeredMessages);
+      }
+    };
 
-        // NOTE: through rhea-promise most of our event handlers are made asynchronous by calling setTimeout(emit).
-        // However, a small set (*error and drain) execute immediately. This can lead to a situation where the logical
-        // ordering of events is correct but the execution order is incorrect because the events are not all getting
-        // put into the task queue the same way.
-        // So this call, while odd, just ensures that we resolve _after_ any already-queued onMessage handlers that may
-        // be waiting in the task queue.
-        setTimeout(() => {
-          cleanupBeforeResolveOrReject();
-          resolve(brokeredMessages);
-        });
-      };
-
-      cleanupBeforeResolveOrReject = (): void => {
-        if (receiver != null) {
-          receiver.removeListener(ReceiverEvents.receiverError, onError);
-          receiver.removeListener(ReceiverEvents.message, onReceiveMessage);
-          receiver.session.removeListener(SessionEvents.sessionError, onError);
-          receiver.removeListener(ReceiverEvents.receiverClose, onClose);
-          receiver.session.removeListener(SessionEvents.sessionClose, onClose);
-          receiver.removeListener(ReceiverEvents.receiverDrained, onReceiveDrain);
+    // Action to be performed on the "message" event.
+    const onReceiveMessage: OnAmqpEventAsPromise = async (context: EventContext) => {
+      // TODO: this appears to be aggravating a bug that we need to look into more deeply.
```

Comment:
Is this TODO still applicable after the changes you've made in the past couple months?

Reply:
I'd like to test this some more but I believe we've nailed the root cause at this point.

Issue filed here just to track that:
https://github.com/Azure/azure-sdk-for-js/issues/13094